The Gaedcke gene expression data was accessed through Gene Expression Omnibus (GEO) using their R package `GEOquery` and its unique id [GSE20842](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE20842). These data were generated on the Agilent-014850 Whole Human Genome Microarray 4x44K G4112F and were reported as being previously quantile normalized using the `limma` package in R before being uploaded to GEO. We extracted the subset of samples which were designated as tumor tissue and removed probes for which there was no gene symbol associated. The first Singular Value was then used to collapse multiple probe to a single expression value per gene where necessary. Singular Value Decomposition (SVD) was performed on the entire resulting expression matrix to assess its structure, as seen below.

```{r loadData, include=FALSE}
## SCRIPT TO QC AND PROCESS THE GAEDCKE DATASET
#####
## ANALYST: BRIAN M. BOT
#####

require(synapseClient)
require(rGithubClient)
require(GEOquery)
require(Biobase)
require(limma)
require(lattice)

## SOURCE IN UTILITY FUNCTIONS
myRepo <- getRepo(repository="Sage-Bionetworks/CRC-RASness",
                  ref="branch", refName="dev")
sourceRepoFile(myRepo, "functions/utilityFunctions.R")

## CREATE A DIRECTORY FOR PLOTS TO BE UPLOADED TO SYNAPSE
gaedckeDir <- file.path(tempdir(), "gaedckeQC")
dir.create(gaedckeDir)

geoFiles <- getGEO("GSE20842", GSEMatrix=T, AnnotGPL=T)
gaedckeEset <- geoFiles$GSE20842_series_matrix.txt.gz
# NOTE: KRAS STATUS - pData(gaedckeEset)$characteristics_ch1.5 IN pData

## SUBSET TO ONLY TUMORS
clin <- pData(gaedckeEset)
gaedckeEset <- gaedckeEset[, clin$characteristics_ch1.4 == "tissue: tumor" ]

## PULL IN THE DATA AND SUBSET TO COHORT OF INTEREST (TUMOR SAMPLES)
exprSet <- gaedckeEset
expr <- exprs(exprSet)
clin <- pData(exprSet)
feat <- fData(exprSet)

## FIX UP CLINICAL DATA A BIT
charNames <- "characteristics_ch1"
charNames <- c(charNames, paste(charNames, 1:5, sep="."))
for(header in charNames){
  tmp <- t(sapply(as.character(clin[[header]]), function(x){ strsplit(x, ": ")[[1]] } ))
  stopifnot(length(unique(tmp[, 1])) == 1)
  name <- unique(tmp[, 1])
  clin[ name ] <- tmp[, 2]
}
clin$kras_status <- ifelse(clin$"genome/variation" == "mutated KRAS", "MUT", "WT")


## GET RID OF GENES FOR WHICH WE HAVE NO ANNOTATION
mapDat <- as.character(feat$"Gene symbol")
mapDat[mapDat == ""] = NA
expr <- expr[ !is.na(mapDat), ]
feat <- feat[!is.na(mapDat), ]
mapDat <- mapDat[ !is.na(mapDat) ]

## COMBINE TO ONE MEASUREMENT PER GENE
exprGene <- combineProbesToGene(expr, mapDat)

## LOOK AT SVD
s <- fs(exprGene)
```



```{r svdRaw, echo=FALSE}
xyplot(s$d ~ 1:length(s$d),
       xaxt="n",
       xlab="eigen gene",
       ylab="% variance explained")
xyplot(s$v[,2] ~ s$v[,1],
       xlab="1st svd",
       ylab="2nd svd")
```

The first two singular values were not strongly associated with any clinical covariates and there appears to be no single outlier nor distinct group of outliers. Age and gender effects were subsequently removed. The following plots are SVD plots after age and gender effects were removed to ensure this did not introduce any unwanted effects to the data.


```{r adjustExpression, include=FALSE}
## ADJUST DATA FOR AGE AND GENDER
fit <- lmFit(exprGene, model.matrix(~as.numeric(clin$age) + factor(clin$gender)))
exprAdj <- residuals(fit, exprGene)

exprSetAdj <- ExpressionSet(exprAdj)
pData(exprSetAdj) <- clin

## PLOTS FOR ADJUSTED EXPRESSION
sAdj <- fs(exprAdj)
```

```{r svdAdjusted, echo=FALSE}
xyplot(sAdj$d ~ 1:length(sAdj$d),
       xaxt="n",
       xlab="eigen gene",
       ylab="% variance explained")
xyplot(sAdj$v[,2] ~ sAdj$v[,1],
       xlab="1st svd",
       ylab="2nd svd")
```

```{r saveObject, include=FALSE}
gaedckeQcedExpressionSet <- exprSetAdj
save("gaedckeQcedExpressionSet", file=file.path(tempdir(), "gaedckeQcedExpressionSet.rda"))
```

This resulting R object is an `ExpressionSet` containing gene-level expression values as well as clinical information on all samples within the `phenoData` slot of the object.

